addons/binding/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/discovery/OpenSprinklerDiscoveryThread.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.opensprinkler.discovery;
+
+import static org.openhab.binding.opensprinkler.OpenSprinklerBindingConstants.DISCOVERY_DEFAULT_IP_TIMEOUT_RATE;
+import static org.openhab.binding.opensprinkler.internal.api.OpenSprinklerApiConstants.*;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+
+import org.openhab.binding.opensprinkler.internal.api.OpenSprinklerApi;
+import org.openhab.binding.opensprinkler.internal.api.OpenSprinklerApiFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link OpenSprinklerDiscoveryThread} class allow manual discovery of
+ * OpenSprinkler devices for OpenHab for a single IP address. This is used
+ * for threading to make discovery faster.
+ *
+ * @author Chris Graham - Initial contribution
+ */
+public class OpenSprinklerDiscoveryThread implements Runnable {
+    private final Logger logger = LoggerFactory.getLogger(OpenSprinklerDiscoveryThread.class);
+
+    private OpenSprinklerDiscoveryService discoveryClass;
+
+    private String ipAddress;
+
+    public OpenSprinklerDiscoveryThread(OpenSprinklerDiscoveryService service, String ip) {
+        this.discoveryClass = service;
+        this.ipAddress = ip;
+    }
+
+    @Override
+    public void run() {
+        if (hasOpenSprinklerDevice(this.ipAddress)) {
+            discoveryClass.submitDiscoveryResults(this.ipAddress);
+        }
+    }
+
+    /**
+     * Determines if an OpenSprinkler device is available at a given IP address.
+     *
+     * @param ip IP address of the OpenSprinkler device as a string.
+     * @return True if a device is found, false if not.
+     */
+    private boolean hasOpenSprinklerDevice(String ip) {
+        try {
+            InetAddress address = InetAddress.getByName(ip);
+
+            if (canEstablishConnection(address, DEFAULT_API_PORT)) {
+                OpenSprinklerApi openSprinkler = OpenSprinklerApiFactory.getHttpApi(ip, DEFAULT_API_PORT,
+                        DEFAULT_ADMIN_PASSWORD);
+
+                return (openSprinkler != null);
+            } else {
+                logger.debug("No OpenSprinkler device found at IP address (" + ip + ")");
```

Comment:
btw, please go through your logging statements and replace string concatenation by parameters, here e.g.
```
logger.debug("No OpenSprinkler device found at IP address ({})", ip);
```